srk/src/terminationLLRF.ml

```diff
@@ -0,0 +1,425 @@
+open Syntax
+module Vec = Linear.QQVector
+module Mat = Linear.QQMatrix
+module TF = TransitionFormula
+
+include Log.Make(struct let name = "TerminationLLRF" end)
+
+type analysis_result = ProvedToTerminate | Unknown
+
+(** Transform a formula representation of a convex set f into a list of linear 
+    inequalities. The linear inequalities are vectors where the meaning
+    of the dimensions are given by the coordinate system cs.
+*)
+let get_polyhedron_of_formula srk f cs =
+  let f = match Formula.destruct srk f with
+    | `And xs -> xs
+    | `Tru -> []
+    | `Atom _ -> [ f ]
+    | _ -> failwith "formula is not convex polyhedron"
+  in
+  Polyhedron.of_implicant ~admit:true cs f
+  |> Polyhedron.enum
+  |> BatList.of_enum
+
+(** From the vector representation of a linear inequality, get the coefficient
+    of a symbol.
+ *)
+let get_coeff_of_symbol cs vec symbol =
+  try 
+    let tid = CoordinateSystem.cs_term_id cs (`App (symbol, [])) in
+    Vec.coeff tid vec
+  with Not_found -> QQ.zero
+
+(** Extracting linear terms that are non-increasing through one iteration.
+    Let X, X' be the sets of pre- and post-transition variables. 
+    Assume we have a convex polyhedron where variables range over X and X',
+    according to Farkas' lemma, every function of the form 
+    a(x - x') + b(y - y') >= 0 must be non-negative combinations of the 
+    half-spaces that consitute the boundaries of the polyhedron.
+    cs: coordinate system
+    ineqs: inequalities as list of vectors
+    dx_list: list of delta symbols defined using dx = x' - x for every x
+    coeff_x_list: a list of symbolic coefficients in the final linear terms
+    coeff_x_set: set of symbolic coefficients
+*)
+let build_system_of_ineq_for_non_inc srk cs ineqs dx_list coeff_x_list coeff_x_set =
```

Comment:
Lots of parameters here, and I don't know how to differentiate this spec from the one for compute_non_inc_cone.  Seems that they should be combined and simplified.